src/librustc_mir/borrow_check.rs

```diff
@@ -846,6 +850,199 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
             }
         }
     }
+
+    /// Check the permissions for the given lvalue and read or write kind
+    fn check_access_permissions(&self, (lvalue, span): (&Lvalue<'tcx>, Span), kind: ReadOrWrite) {
+        match kind {
+            Write(WriteKind::MutableBorrow(BorrowKind::Unique))
+            | Write(WriteKind::MutableBorrow(BorrowKind::Mut)) => {
+                if let Err(_lvalue_err) = self.is_mutable(lvalue) {
+                    self.tcx.cannot_borrow_path_as_mutable(span,
+                        &format!("immutable variable `{}`",
+                                  self.describe_lvalue(lvalue)),
+                        Origin::Mir)
+                        .emit();
+                }
+            },
+            Write(WriteKind::Move) => {
+                if let Err(_lvalue_err) = self.is_mobile(lvalue) {
+                    self.tcx.cannot_borrow_path_as_mutable(span,
+                        &format!("immovable variable `{}`",
+                                  self.describe_lvalue(lvalue)),
+                        Origin::Mir)
+                        .emit();
+                }
+            },
+            _ => {}// Access authorized
+        }
+    }
+
+    /// Can this value be written or borrowed mutably
+    fn is_mutable<'d>(&self, lvalue: &'d Lvalue<'tcx>) -> Result<(), &'d Lvalue<'tcx>> {
+        match *lvalue {
+            Lvalue::Local(local) => {
+                let local = &self.mir.local_decls[local];
+                if local.mutability == Mutability::Not {
```

Comment:
I would prefer an (exhaustive) `match` here, which makes it obvious that we've covered every case. I would say that we should keep `if` for booleans.